directembedding/src/main/scala/ch/epfl/directembedding/transformers/ReifyAsEmbedding.scala

```diff
@@ -84,42 +80,93 @@ trait ReifyAsEmbedding extends DirectEmbeddingModule with DirectEmbeddingUtils {
+    private def args(lhs: Tree, rhs: List[Tree]): List[Tree] = {
+      (self(lhs).toList ::: rhs).map(transform)
+    }
 
     private def symbol(lhs: Tree): Symbol = lhs match {
-      case Select(New(newBody), _)            => newBody.symbol
-      case field @ Select(_, _)               => getInvokedSymbol(field)
-      case TypeApply(field @ Select(_, _), _) => getInvokedSymbol(field)
-      case _                                  => lhs.symbol
+      case Select(New(newBody), _)     => newBody.symbol
+      case field: Select               => invokedSymbol(field)
+      case TypeApply(field: Select, _) => invokedSymbol(field)
+      case _                           => lhs.symbol
     }
 
+    private def selfType(tree: Tree): Type = {
+      val owner = tree.symbol.owner
+      if (owner.isType) owner.asType.toType
+      else throw new RuntimeException(s"Unable to find type of $tree")
+    }
+
+    /**
+     * Get body of class to be reified through [[reifyAs]] annotation.
+     *
+     * First, performs lookup for the [[reifyAs]] annotation in the invoked
+     * symbol and falls back to a lookup in [[typeMap]] for the type of
+     * parameter `tree` if no annotation exists.
+     *
+     * @param treeSymbol Symbol of invoked symbol
+     * @param tree Tree to be reified through annotation
+     * @return Reified tree if annotation exists, fails compilation otherwise
+     */
+    private def reificationBody(treeSymbol: Symbol, tree: Tree): Tree = {
+      reifyAnnotation(treeSymbol).orElse {
+        log(s"Missing reifyAs annotation for $treeSymbol")
+        for {
+          fallbackType <- typeMap.get(selfType(tree).typeSymbol.fullName)
+          fallbackSymbol <- {
+            log(s"Falling back to $fallbackType")
+            findMatchingSymbol(treeSymbol, fallbackType)
+          }
+          fallbackAnnotation <- {
+            log(s"Found matching symbol in overridden type: $fallbackSymbol")
+            reifyAnnotation(fallbackSymbol)
+          }
+        } yield fallbackAnnotation
+      }.map {
+        _.tree.children.tail.headOption.getOrElse {
+          throw new IllegalArgumentException("Missing argument for reifyAs annotation")
+        }
+      }
+    }.getOrElse {
+      c.abort(pos, s"$treeSymbol is not supported in $dslName")
+    }
+
+    private def body(lhs: Tree, tree: Tree): Tree =
+      reificationBody(symbol(lhs), tree)
+
+    private var indent: Int = 0
+
     override def transform(tree: Tree): Tree = {
-      logIndented(s"transform(): $tree", indent, logLevel)
+      logIndented(s"transform(): $tree", indent)
       logTree(tree, logLevel + 1)
       indent += 2
       val result = tree match {
+        // f(a)(b) => f(a, b)
         case a @ Apply(Apply(_, _), _) =>
-          logIndented(s"Apply(Apply)", indent, logLevel)
+          logIndented(s"Apply(Apply)", indent)
           transform(uncurry(a, Nil))
 
+        // (expr: typ) => expr
+        case Typed(expr, typ) =>
+          transform(expr)
```

Comment:
Is this sufficient to solve typed blocks or do we need to use `TypeTransformation` from Yin-Yang?